src/main/scala/firrtl/passes/PadWidths.scala

```diff
@@ -7,63 +7,59 @@ import firrtl.ir._
+    case _                => false
   }
 
-  private def width(t: Type):       Int = bitWidth(t).toInt
-  private def width(e: Expression): Int = width(e.tpe)
-  // Returns an expression with the correct integer width
-  private def fixup(i: Int)(e: Expression) = {
-    def tx = e.tpe match {
-      case t: UIntType => UIntType(IntWidth(i))
-      case t: SIntType => SIntType(IntWidth(i))
-      // default case should never be reached
-    }
-    width(e) match {
-      case j if i > j => DoPrim(Pad, Seq(e), Seq(i), tx)
-      case j if i < j =>
-        val e2 = DoPrim(Bits, Seq(e), Seq(i - 1, 0), UIntType(IntWidth(i)))
-        // Bit Select always returns UInt, cast if selecting from SInt
-        e.tpe match {
-          case UIntType(_) => e2
-          case SIntType(_) => DoPrim(AsSInt, Seq(e2), Seq.empty, SIntType(IntWidth(i)))
-        }
-      case _ => e
+  /** Adds padding or a bit extract to ensure that the expression is of the with specified.
+    * @note only works on UInt and SInt type expressions, other expressions will yield a match error
+    */
+  private[firrtl] def forceWidth(width: Int)(e: Expression): Expression = {
+    val old = getWidth(e)
+    if (width == old) { e }
+    else if (width > old) {
+      // padding retains the signedness
+      val newType = e.tpe match {
+        case _: UIntType => UIntType(IntWidth(width))
+        case _: SIntType => SIntType(IntWidth(width))
+        case other => throw new RuntimeException(s"forceWidth does not support expressions of type $other")
+      }
+      ConstantPropagation.constPropPad(DoPrim(Pad, Seq(e), Seq(width), newType))
+    } else {
+      val extract = DoPrim(Bits, Seq(e), Seq(width - 1, 0), UIntType(IntWidth(width)))
+      val e2 = ConstantPropagation.constPropBitExtract(extract)
+      // Bit Select always returns UInt, cast if selecting from SInt
+      e.tpe match {
+        case UIntType(_) => e2
+        case SIntType(_) => DoPrim(AsSInt, Seq(e2), Seq.empty, SIntType(IntWidth(width)))
+      }
     }
   }
 
+  private def getWidth(t: Type):       Int = bitWidth(t).toInt
+  private def getWidth(e: Expression): Int = getWidth(e.tpe)
+
   // Recursive, updates expression so children exp's have correct widths
   private def onExp(e: Expression): Expression = e.map(onExp) match {
     case Mux(cond, tval, fval, tpe) =>
-      Mux(cond, fixup(width(tpe))(tval), fixup(width(tpe))(fval), tpe)
-    case ex: ValidIf => ex.copy(value = fixup(width(ex.tpe))(ex.value))
+      Mux(cond, forceWidth(getWidth(tpe))(tval), forceWidth(getWidth(tpe))(fval), tpe)
+    case ex: ValidIf => ex.copy(value = forceWidth(getWidth(ex.tpe))(ex.value))
     case ex: DoPrim =>
       ex.op match {
-        case Lt | Leq | Gt | Geq | Eq | Neq | Not | And | Or | Xor | Add | Sub | Rem | Shr =>
+        case Lt | Leq | Gt | Geq | Eq | Neq | And | Or | Xor | Add | Sub =>
           // sensitive ops
```

Comment:
`Not`: input width == output width, so there is nothing to pad
![Screenshot from 2021-08-02 13-22-17](https://user-images.githubusercontent.com/2456545/127919076-935eeab5-2ef3-4b75-82f6-b7f13d81c1c6.png)

Reply:
`Rem`: widening the arguments does not really make sense since the result with is calculated with `min`
![Screenshot from 2021-08-02 13-24-38](https://user-images.githubusercontent.com/2456545/127919407-1cfc6bc4-7761-45c0-b5db-14f02123a027.png)

Reply:
`Shr`: if we widen the argument, it will widen the output width
![Screenshot from 2021-08-02 13-26-04](https://user-images.githubusercontent.com/2456545/127919515-d882cbea-e3d6-42fc-8338-92055263b637.png)

Reply:
Or to summarize, the only ops, where we can do this padding without violating firrtl width semantics are the ones where the result width is the `max` of the input widths (+ const)

Reply:
I will add this as a comment

Reply:
Awesome, thanks for the explanation!